Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+                    // since the closure is guaranteed to be run straight away
+                    return await withCheckedContinuation { continuation in
+                        let action = self.stateMachine.next(for: continuation)
+
+                        switch action {
+                        case .callProduceMore:
+                            let delegate = delegate
+                            self.lock.unlock()
+
+                            delegate?.produceMore()
+
+                        case .none:
+                            self.lock.unlock()
+                        }
+                    }
+                }
+            } onCancel: {
+                let delegate: Delegate? = self.lock.withLock {
+                    let action = self.stateMachine.cancelled()
+
+                    switch action {
+                    case .callDidTerminate:
+                        let delegate = self.delegate
+                        self.delegate = nil
+
+                        return delegate
+
+                    case .resumeContinuationWithNilAndCallDidTerminate(let continuation):
+                        continuation.resume(returning: nil)
+                        let delegate = self.delegate
+                        self.delegate = nil
+
+                        return delegate
+
+                    case .none:
+                        return nil
+                    }
+                }
+
+                delegate?.didTerminate()
+            }
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncSequenceProducer {
+    @usableFromInline
+    /* private */ internal struct StateMachine {
+        @usableFromInline
+        /* private */ internal enum State {
+            /// The initial state before either a call to `yield()` or a call to `next()` happened
+            case initial(
+                backPressureStrategy: Strategy,
+                iteratorInitialized: Bool
+            )
+
+            /// The state once either any element was yielded or `next()` was called.
+            case streaming(
+                backPressureStrategy: Strategy,
+                buffer: Deque<Element>,
+                continuation: CheckedContinuation<Element?, Never>?,
+                hasOutstandingDemand: Bool,
+                iteratorInitialized: Bool
+            )
+
+            /// The state once the underlying source signalled that it is finished.
+            case sourceFinished(
+                buffer: Deque<Element>,
+                iteratorInitialized: Bool
+            )
+
+            /// The state once there can be no outstanding demand. This can happen if:
+            /// 1. The ``NIOAsyncSequenceProducer/AsyncIterator`` was deinited
+            /// 2. The underlying source finished and all buffered elements have been consumed
+            case finished
+
+            /// Internal state to avoid CoW.
+            case modifying
```

Comment:
I quickly tried to create a reproducer for this and you are right that in a simple case the compiler is smart enough to avoid the CoW here. I have also quickly talked with @Lukasa and he explained me that in more complicated cases, especially if `@inlinable` methods are in play the compiler is not always able to optimise this away.

Since the code is already there and it helps the compiler, I would prefer to keep it.